### PR TITLE
Lab 19

### PR DESCRIPTION
**Pull request by oreny**

### CODE REVIEW

File: src/sorted-table.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
+import React from "react";
+import {useState} from "react";
+import "./sorted-table.css";
+
+export default function SortedTable({tableData}) {
+
+    const tableContentWithKey = () => tableData.slice(1).map((row, i) => [i].concat(row));
+    const compare = (a,b) =>
+        typeof a === "number" ? a - b
+            : typeof a === "string" ? (a.toLowerCase() < b.toLowerCase() ? -1 : a.toLowerCase() > b.toLowerCase() ? 1 : 0)
```
Comment (ilyador) [inline src/sorted-table.js]: Why do you test for string/boolean?
It is hard to read because all your sorting functions use `a` and `b` as variable names.

File: src/sorted-table.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        return tbl.map((row, i) => [row[column], i])
+            .sort((a, b) => compare(a[0], b[0]) * (asc ? 1 : -1))
+            .map(a => tbl[a[1]]);
+    }
+
+    function onSortClick(columnIndex) {
+        const asc = columnIndex !== sortColumn || !ascending;
+        setSortColumn(() => columnIndex);
+        setTableContent(sortTable(tableContent, columnIndex, asc));
+        setAscending(() => asc);
```
Comment (ilyador) [inline src/sorted-table.js]: Why are you using a function here?

File: src/sorted-table.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+    function sortTable(tbl, column, asc) {
+        return tbl.map((row, i) => [row[column], i])
+            .sort((a, b) => compare(a[0], b[0]) * (asc ? 1 : -1))
+            .map(a => tbl[a[1]]);
+    }
+
+    function onSortClick(columnIndex) {
+        const asc = columnIndex !== sortColumn || !ascending;
+        setSortColumn(() => columnIndex);
```
Comment (ilyador) [inline src/sorted-table.js]: Why are you using a function here?

File: src/sorted-table.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+    const [sortColumn, setSortColumn] = useState(0);
+    const [tableContent, setTableContent] = useState(tableContentWithKey());
+    const [ascending, setAscending] = useState(true);
+
+
+
+    function sortTable(tbl, column, asc) {
+        return tbl.map((row, i) => [row[column], i])
+            .sort((a, b) => compare(a[0], b[0]) * (asc ? 1 : -1))
```
Comment (ilyador) [inline src/sorted-table.js]: I got lost here. I think this is overcomplicated and quite hard to follow the logic.
I would suggest to think of a simpler way to do it.

File: src/sorted-table.js (hunk 1 above)
Comment (oreny) [reply src/sorted-table.js]: How would you name two variables of a compare function?

File: src/sorted-table.js (hunk 2 above)
Comment (oreny) [reply src/sorted-table.js]: According to Ynon, it is better to provide to provide a function and not the actual value.
Let me know if there's a refinement to that guideline

File: src/sorted-table.js (hunk 3 above)
Comment (oreny) [reply src/sorted-table.js]: According to Ynon, it is better to provide to provide a function and not the actual value.
Let me know if there's a refinement to that guideline

Comment (ilyador) [reply src/sorted-table.js]: I'm not sure what @ynonp meant by that, but the reason to use a function is if you want access to the current value.
If you simply pass a value, there is no need for a function.

File: src/sorted-table.js (hunk 1 above)
Comment (ilyador) [reply src/sorted-table.js]: Call them by what it is you are comparing, `tableRow1` for example

Comment (oreny) [reply src/sorted-table.js]: I do not agree in this case. It is a generic comparison function that compares two parameters.
Maybe it should reside elsewhere. I moved it outside of the component so maybe it is clearer now;

File: src/sorted-table.js (hunk 4 above)
Comment (oreny) [reply src/sorted-table.js]: You're right. I modified a bit to make it more readable

File: src/sorted-table.js (hunk 2 above)
Comment (ynonp) [reply src/sorted-table.js]: Just saw this - the guideline is: 

1. Use a function when the change *uses* the previous value (for example increase a value by one)

2. Use a fixed value when the change is not about changing the previous value (for example "reset" which sets always to 0)

File: src/sorted-table.js (hunk 1 above)
Comment (ilyador) [reply src/sorted-table.js]: This function is not generic at all. You use it for something very specific in this project. You can't just copy paste it into another without modifications.

There will be very rare cases that you will actually use a compare function so generic that it can take any possible value.

Moreover, the idea of using a function so generic to solve a specific issue with specific data, is the wrong approach. Never try to solve problems that don't yet exist.

In this case you try to make it more generic by checking for numbers which simply do not exist in the data.
